themis/spec/asymmetric-keypairs/rsa.md

```diff
@@ -11,3 +11,442 @@ Please come back later.
+b4TlUDlsGyKj9ESG7zuBO9UM2yzquLMnf+6yrRgjUZ+n2EuAdg6bOhIESm1a+vC13mi7HOpkwz/K
+ZJpmWztsuouOncUy21tWwtGJafX2UkeQPxiN8G/4LA/thiar9OYX3h5Z4lEFlapiznPPRc1mLzxQ
+uwNHsii+9HU5uVZzQnXsU3Fw52zTqDzcRfkqKT+YBi+RBHbldNkhnZ4ULlZ4cvSEzMbpmD/JCd4d
+jvUf8FJz3XSkN9AYyHLbXSDJqKZnLPJdAAEAAQ==`
+	privateKey, _ := base64.StdEncoding.DecodeString(privateKeyBase64)
+
+	// Extract the values of key parameters
+	d := new(big.Int).SetBytes(privateKey[0x00C:0x10C])
+	p := new(big.Int).SetBytes(privateKey[0x10C:0x18C])
+	q := new(big.Int).SetBytes(privateKey[0x18C:0x20C])
+	dp := new(big.Int).SetBytes(privateKey[0x20C:0x28C])
+	dq := new(big.Int).SetBytes(privateKey[0x28C:0x30C])
+	qp := new(big.Int).SetBytes(privateKey[0x30C:0x38C])
+	n := new(big.Int).SetBytes(privateKey[0x38C:0x48C])
+	e := new(big.Int).SetBytes(privateKey[0x48C:0x490])
+
+	// p and q must be primes (well, with a 1 − 1/4^25 confidence)
+	if p.ProbablyPrime(25) && q.ProbablyPrime(25) {
+		fmt.Println("✅ p and q are primes")
+	} else {
+		fmt.Println("❌ p and q are primes")
+	}
+	// ...and by definition, n = p⋅q
+	nComputed := new(big.Int).Mul(p, q)
+	if nComputed.Cmp(n) == 0 {
+		fmt.Println("✅ n = p⋅q")
+	} else {
+		fmt.Println("❌ n = p⋅q")
+	}
+
+	pMinusOne := new(big.Int).Sub(p, big.NewInt(1))
+	qMinusOne := new(big.Int).Sub(q, big.NewInt(1))
+
+	// By definition, λ(n) = λ(p⋅q) = LCM(λ(p), λ(q)) =
+	//   = LCM(p−1, q−1) = (p−1)(q−1) / GCD(p−1, q−1)
+	pq1Mul := new(big.Int).Mul(pMinusOne, qMinusOne)
+	pq1GCD := new(big.Int).GCD(nil, nil, pMinusOne, qMinusOne)
+	totient := new(big.Int).Div(pq1Mul, pq1GCD)
+
+	// e must be within 1 < e < λ(n)
+	if (big.NewInt(1).Cmp(e) == -1) && (e.Cmp(totient) == -1) {
+		fmt.Println("✅ 1 < e < λ(n)")
+	} else {
+		fmt.Println("❌ 1 < e < λ(n)")
+	}
+	// ...plus, e and λ(n) must be coprime (i.e., GCD(e, λ(n)) = 1)
+	gcd := new(big.Int).GCD(nil, nil, e, totient)
+	if gcd.Cmp(big.NewInt(1)) == 0 {
+		fmt.Println("✅ e is coprime with λ(n)")
+	} else {
+		fmt.Println("❌ e is coprime with λ(n)")
+	}
+
+	// It must be that d⋅e ≡ 1 (mod λ(n))
+	dXe := new(big.Int).Mod(new(big.Int).Mul(d, e), totient)
+	if dXe.Cmp(big.NewInt(1)) == 0 {
+		fmt.Println("✅ d⋅e ≡ 1 (mod λ(n))")
+	} else {
+		fmt.Println("❌ d⋅e ≡ 1 (mod λ(n))")
+	}
+
+	// By definition, dp ≡ d (mod p−1)
+	dpComputed := new(big.Int).Mod(d, pMinusOne)
+	if dpComputed.Cmp(dp) == 0 {
+		fmt.Println("✅ dp ≡ d (mod p−1)")
+	} else {
+		fmt.Println("❌ dp ≡ d (mod p−1)")
+	}
+
+	// By definition, dq ≡ d (mod q−1)
+	dqComputed := new(big.Int).Mod(d, qMinusOne)
+	if dqComputed.Cmp(dq) == 0 {
+		fmt.Println("✅ dq ≡ d (mod q−1)")
+	} else {
+		fmt.Println("❌ dq ≡ d (mod q−1)")
+	}
+
+	// It must be that qp⋅q ≡ 1 (mod p)
+	qpXq := new(big.Int).Mod(new(big.Int).Mul(qp, q), p)
+	if qpXq.Cmp(big.NewInt(1)) == 0 {
+		fmt.Println("✅ qp⋅q ≡ 1 (mod p)")
+	} else {
+		fmt.Println("❌ qp⋅q ≡ 1 (mod p)")
+	}
+}
+```
+
+## Generating RSA keys
+
+The recommended way to generate RSA keypairs is to use a cryptography library.
```

Comment:
seems it should be "cryptographic library"

Reply:
> seems it should be "cryptographic library"

IDK :joy: During editing I've flipped it between "cryptography" and "cryptographic" several times. Unless there are strong opinions, I guess I'll go ahead to grep the variants in our own docs and just use the most popular one.

Reply:
"Cryptographic" it is.

```
$ git grep -hoi 'crypto[a-z]*' | sort | uniq -c
    351 crypto
      3 Crypto
      5 CRYPTO
      2 cryptoalgorithm
      5 cryptoalgorithms
      1 cryptoanalysis
      2 cryptographer
      3 cryptographers
    135 cryptographic
      4 Cryptographic
     12 cryptographically
     90 cryptography
     15 Cryptography
      1 cryptograpic
      1 cryptomuseum
      5 cryptoperiod
      1 cryptoperiods
     81 cryptosystem
      3 Cryptosystem
    311 cryptosystems
      5 Cryptosystems
```

Reply:
`cryptography library` is more common used term, better for search

https://trends.google.com/trends/explore?q=cryptography%20library,cryptographic%20library
<img width="288" alt="Screenshot 2020-09-11 at 14 00 15" src="https://user-images.githubusercontent.com/2877920/92917418-23d8a280-f437-11ea-9b22-86c88041ca0a.png">

Reply:
> cryptography library is more common used term, better for search

Okie dokie, reverted.